### PR TITLE
Switch synchronous thread to thread_local storage

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

IOCTLs sent via a file handle that is opened without the FILE_FLAG_OVERLAPPED flag are serialized in the file object. This will affect performance if multiple user mode threads are making concurrent IOCTL requests.

This change switches the synchronous handle to being per-thread, with the handle being closed on thread detach.

## Testing

CI/CD

## Documentation

No.

## Installation

No.

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
Need to investigate failures.

**Comment by Alan-Jowett** [comment]
Isusue resolved, ready for review.

**Comment by matthewige** [comment]
Code itself looks good to me, but do we have existing tests which utilize ebpf api on multiple threads?

File: ebpfapi/dllmain.cpp [hunk 1]
```diff
@@ -21,8 +21,10 @@ DllMain(HMODULE hModule, unsigned long ul_reason_for_call, void* lpReserved)
         }
         break;
     case DLL_THREAD_ATTACH:
+        ebpf_api_thread_local_initialize();
         break;
     case DLL_THREAD_DETACH:
+        ebpf_api_thread_local_cleanup();
```
Comment (matthewige) [inline ebpfapi/dllmain.cpp:27]: Is it possible that DLL_PROCESS_DETACH gets trigged but not DLL_THREAD_DETACH? In other words, do we ever care about invoking ebpf_api_thread_local_cleanup in DLL_THREAD_DETACH?

Comment (Alan-Jowett) [reply ebpfapi/dllmain.cpp:27]: DLL_PROCESS_DETACH means the process is being terminated. All resources associated with the process will be terminated, so cleaning up handles is moot.

**Comment by Alan-Jowett** [comment]
> Code itself looks good to me, but do we have existing tests which utilize ebpf api on multiple threads?

Added dedicates IOCTL stress test, FWIW.

**Review by matthewige — APPROVED** [review]

File: ebpfapi/dllmain.cpp (hunk 1 above)
Comment (saxena-anurag) [inline ebpfapi/dllmain.cpp:27]: question: Is `DLL_THREAD_DETACH` guaranteed to be called for all the existing threads when the DLL is been unloaded?

Comment (Alan-Jowett) [reply ebpfapi/dllmain.cpp:27]: This is what the MSDN says:
<html>
<body>

DLL_THREAD_DETACH3 | A thread is exiting cleanly. If the DLL has stored a pointer to allocated memory in a TLS slot, it should use this opportunity to free the memory. The system calls the entry-point function of all currently loaded DLLs with this value. The call is made in the context of the exiting thread.
-- | --



</body>
</html>

**Comment by gtrevi** [comment]
Would it make sense to implement `_sync_device_handle` and `_async_device_handle` as instances of a "device_handle" class (w/constr(), destr() and get(), eventually with an extra-safety on the thread-detach system call (even though the compiler should guarantee the destructor to be called)?

File: libs/api_common/device_helper.cpp [hunk 2]
```diff
@@ -27,21 +27,35 @@ typedef struct _async_ioctl_completion_context
 } async_ioctl_completion_context_t;
 
 // Handle used for synchronous calls to the driver.
```
Comment (gtrevi) [inline libs/api_common/device_helper.cpp:29]: Would it make sense to implement `_sync_device_handle` and `_async_device_handle` as instances of a new "device_handle" class (w/constr(), destr() and get()), eventually with an extra-safety on the thread-detach system call (even though the compiler should guarantee the destructor to be called)?

File: libs/api_common/device_helper.cpp [hunk 3]
```diff
@@ -27,21 +27,35 @@ typedef struct _async_ioctl_completion_context
...
-static ebpf_handle_t _sync_device_handle = ebpf_handle_invalid;
+static thread_local ebpf_handle_t _sync_device_handle = ebpf_handle_invalid;
```
Comment (gtrevi) [inline libs/api_common/device_helper.cpp:30]: Would it make sense to implement `_sync_device_handle` and `_async_device_handle` as instances of a new "device_handle" class (w/constr(), destr() and get()), eventually with an extra-safety on the thread-detach system call (even though the compiler should guarantee the destructor to be called)?

Comment (gtrevi) [inline libs/api_common/device_helper.cpp:30]: Would it make sense to implement `_sync_device_handle` and `_async_device_handle` as instances of a new `device_handle` class (w/constr(), destr() and get())?

Comment (Alan-Jowett) [reply libs/api_common/device_helper.cpp:30]: I tried this approach at first, but this causes ab order of operations issue with the test code. We need to have the handles closed before the global destructors are called for the CloseHandle mocks in the test code. DLL detach is called before global destructor, so the test code works correctly.

If we can figure out how to sequence thread_local destructor with destruction of static globals, then maybe we could make this work.

Until then, this is the simplest approach that works.

File: tests/api_test/api_test.cpp [hunk 4]
```diff
@@ -1122,4 +1122,110 @@ TEST_CASE("load_native_program_invalid5", "[native][negative]")
...
+                            failure_count++;
+                        }
+                        break;
+                    }
+                };
+            });
+        }
+    }
+
+    // Wait for 10 seconds
```
Comment (Alan-Jowett) [inline tests/api_test/api_test.cpp:1214]: Comment is out of date.

**Pull request merged**